container/credentials.bzl

```diff
@@ -0,0 +1,118 @@
+load("//container:providers.bzl", "RegistryCredentialInfo")
+
+def write_docker_config_json_action(ctx, registry, credential, config_json = None):
+    """Create an action that writes a docker config file.
+
+    Args:
+      ctx: the rule ctx object
+      registry: the name of the registry to which the credential should be
+      associated
+      credential: a RegistryCredentialInfo struct
+      config_json: path to the file where the config.json should be written.
+      Defaults to %{credential.name}/.docker/config.json.
+
+    Return: The config_json file 
+    """
+    if credential.type != "basic":
+        fail("Unsupported docker credential type: %s" % credential.type)
+    if not config_json:
+        config_json = ctx.actions.declare_file("%s/.docker/config.json" % credential.name)
+    # NOTE: ctx.actions.write does not work here as we need the
+    # use_default_shell_env feature of the run_shell action.
+    ctx.actions.run_shell(
+        mnemonic = "WriteDockerConfig",
+        inputs = [],
+        outputs = [config_json],
+        command = """
+set -euo pipefail
+cat <<EOF > "%s"
+{
+    "auths": {
+        "%s": {
+            "username": "${%s}",
+            "password": "${%s}"
+        }
+    }
+}
+EOF
+        """ % (config_json.path, registry, credential.user_env, credential.pass_env),
+        use_default_shell_env = True,
+    )
+    return config_json
+
+
+def _registry_credential_impl(ctx):
+    """Implementation of the registry credential
+
+    Provides the RegistryCredentialInfo object from given rule attributes.
+    """
+    return [RegistryCredentialInfo(
+        name = ctx.label.name,
+        user_env = ctx.attr.user_env,
+        pass_env = ctx.attr.pass_env,
+        type = ctx.attr.type,
+    )]
+
+registry_credential = rule(
```

Comment:
I'm not sure if this should be a repository rule or a skylark rule as it is now as this is a bit problematic.
Specifically, there are two issues:
- this rule will not work for container_pull, as container_pull is a repository rule and cannot access outputs of a skylark rule
- env vars like this one should be read by repository rules, not skylark rules, particularly because if the user changes the values of the env variables, Bazel will not regenerate the outputs unless you do a Bazel clean

However, I'm also not sure this should be a repository_rule. Specifically because:
- repo rules cannot have providers, so the elegant solution that you have to avoid users from checking in their secret values cannot be used here.

Thoughts?

Reply:
You're right about `container_pull` would need a different mechanism.  I'll have to think about how to reconcile this.

Regarding changing of environment variables.  The implementation of `--action_env` includes the env vars into the SkyKey (or whatever the action hash is called), so in fact it does re-run the rule if you change the value of the environment variable.

Reply:
You're right about env vars, did not remember the action key does include those values. 
let me know what you think about container_pull